### PR TITLE
feat: [DHIS2-12362][DHIS2-12455] View and filter relationships

### PR DESCRIPTION
**Pull request by eirikhaugstulen**
This PR builds on the work of @jasminenguyennn and @JoakimSM.

Add a TrackedEntity relationships widget in the enrollment pages.

TL;DR:

- Added self contained widget that fetches relationships and metadata required
- Also able to pass in cached relationshiptypes (if so, no metadata will be fetched in the widget)
- Creates portal to ref when adding new relationship

### CODE REVIEW

File: src/core_modules/capture-core/components/WidgetsRelationship/common/hooks/useLinkedEntityGroups.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
+// @flow
+import { useCallback, useEffect, useState } from 'react';
+import log from 'loglevel';
+import moment from 'moment';
+import { errorCreator } from 'capture-core-utils';
```
Comment (eirikhaugstulen) [inline src/core_modules/capture-core/components/WidgetsRelationship/common/hooks/useLinkedEntityGroups.js:5]: Can we use deps from our internal repo?

File: src/core_modules/capture-core/reducers/descriptions/organisationUnits.reducerDescription.js [hunk 2]
```diff
@@ -42,5 +42,5 @@ export const organisationUnitRootsDesc = createReducerDescription({
             roots: action.payload.roots,
         },
     }),
-}, 'organisationUnitRoots');
+}, 'organisationUnitRoots2222');
```
Comment (eirikhaugstulen) [inline src/core_modules/capture-core/reducers/descriptions/organisationUnits.reducerDescription.js:45]: Was this introduced by something you thought of @JoakimSM? Do we keep it?

File: src/core_modules/capture-core/utils/reactQueryHelpers/query/useApiDataQuery.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        queryKey,
+        queryFn,
+        {
+            ...queryOptions,
+            refetchOnWindowFocus: false,
+            refetchOnMount: false,
+            refetchOnReconnect: false,
+            // TODO: Not sure what we want here either
+            staleTime: 4 * 60 * 1000,
+            cacheTime: 10 * 60 * 1000,
```
Comment (eirikhaugstulen) [inline src/core_modules/capture-core/utils/reactQueryHelpers/query/useApiDataQuery.js]: What do we think is a sensible amount here?

File: src/core_modules/capture-core/utils/reactQueryHelpers/query/useMetadataQuery.js [hunk 4]
```diff
@@ -38,3 +50,19 @@ export const useIndexedDBQuery = <TResultData>(
...
+    queryObject: ResourceQuery,
+    queryOptions?: UseQueryOptions<TResultData>,
+): Result<TResultData> => {
+    const dataEngine = useDataEngine();
+    const queryFn: QueryFunction<TResultData> = () => dataEngine.query({ theQuerykey: queryObject })
+        .then(response => response.theQuerykey);
+    return useAsyncMetadata(queryKey, queryFn, {
+        // TODO: What do you think is sensible here?
+        cacheTime: Infinity,
+        staleTime: Infinity,
```
Comment (eirikhaugstulen) [inline src/core_modules/capture-core/utils/reactQueryHelpers/query/useMetadataQuery.js:64]: When do we want to refetch metadata from the API?

File: src/core_modules/capture-core/components/WidgetsRelationship/WidgetTrackedEntityRelationship/NewTrackedEntityRelationship/Breadcrumbs/Breadcrumbs.component.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+const FindExistingStep = ({ currentStep }) => {
+    if (currentStep.value < NEW_TRACKED_ENTITY_RELATIONSHIP_WIZARD_STEPS.FIND_EXISTING_LINKED_ENTITY.value) {
+        return null;
+    }
+
+    return (
+        <>
+            <Slash />
+            <span>{'Search'}</span>
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/WidgetsRelationship/WidgetTrackedEntityRelationship/NewTrackedEntityRelationship/Breadcrumbs/Breadcrumbs.component.js:48]: Missing translation?

File: src/core_modules/capture-core/components/WidgetsRelationship/common/hooks/useLinkedEntityGroups.js (hunk 1 above)
Comment (simonadomnisoru) [reply src/core_modules/capture-core/components/WidgetsRelationship/common/hooks/useLinkedEntityGroups.js:5]: `errorCreator` is already used by other self-contained widgets. It looks like an easy function to re-rewrite, so maybe we can move it inside each widget?

File: src/core_modules/capture-core/utils/reactQueryHelpers/query/useMetadataQuery.js (hunk 4 above)
Comment (JoakimSM) [reply src/core_modules/capture-core/utils/reactQueryHelpers/query/useMetadataQuery.js:64]: I think `Infinity` is fine.

File: src/core_modules/capture-core/components/WidgetsRelationship/common/hooks/useRelationshipTypes.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+    const { attributeQuery, dataElementQuery } = useMemo(() => {
+        if (!apiRelationshipTypes) return {};
+        const { dataElementIds, attributeIds } = extractElementIdsFromRelationshipTypes(apiRelationshipTypes);
+
+        const filteredAttributeQuery = {
+            resource: 'trackedEntityAttributes',
+            params: {
+                fields: 'id,displayName,valueType',
+                filter: `id:in:[${Object.keys(attributeIds).join(',')}]`,
```
[the remark this answers is not in the document]
Comment (eirikhaugstulen) [reply src/core_modules/capture-core/components/WidgetsRelationship/common/hooks/useRelationshipTypes.js:38]: Not sure I completely understood what you meant here, @JoakimSM. Let's have a chat the next time we speak.

**Review by simonadomnisoru — COMMENTED** [review]
Hey @eirikhaugstulen,
I get `TypeError: trackedEntityType.name is undefined` when clicking the `New Relationship` button. Can you have a look?

https://github.com/dhis2/capture-app/assets/6817073/baa45461-049e-4ec6-ae51-9c286fbf72f9

Thanks!

File: src/core_modules/capture-core/components/WidgetsRelationship/common/Types/index.js
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/WidgetsRelationship/common/Types/index.js:1]: Should the folder `Types` name start with lowercase instead?

File: src/core_modules/capture-core/components/WidgetsRelationship/common/RelationshipsWidget/linkedEntitiesViewer.types.js
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/WidgetsRelationship/common/RelationshipsWidget/linkedEntitiesViewer.types.js:1]: Does it make sense to move this file, as well as the rest of the `*.types.js` from this folder to the `RelationshipsWidget/types` folder? Thank you!

**Pull request merged**